File: codingChallenges/cpp/Codechef/longChallenges/Year2020/may/bitwise.py
t = int(input())

def getBinary(num):
    ans = []
    while(num > 0):
        ans.append(num%2)
        num=num//2
    return ans[::-1]

def getDecimal(num):
    sum = 0
    for i in num:
        sum*=2
        sum += i
    return sum

def IsFirstGreaterThanSecond(num1,num2):
    if len(num1) > len(num2):
        return True
    elif len(num2) >len(num1):
        return False
    
    for i in range(len(num1)):
        if num1[i] == num2[i]:
            continue
        elif num1[i] == 1:
            return True
        return False
    return False
def cal(a, b, c):
    if (a & c) == 0 or (b & c) == 0 :
        print(0)
        return
    print(c)

def fValue(a,b,c):
    return (a & c) * (b & c)

while(t):
    t-=1
    a,b,l,r =[int(i) for i in input().split(" ")]

    c = a | b

    if c <= r:
        cal(a,b,c)
        continue

    
    binC = getBinary(c)
    binR = getBinary(r)

    sizeDiff = len(binC) - len(binR)
    if sizeDiff > 0 :
        binC = binC[sizeDiff:]

    decC = getDecimal(binC)
    decR = getDecimal(binR)
    if IsFirstGreaterThanSecond(binR, binC):
        cal(a,b,decC)
        continue

    tempAns = decC & decR
    binC.pop(0)

    decC = getDecimal(binC)
    if fValue(a,b,tempAns) > fValue(a,b,decC):
        print(tempAns)
        continue

    cal(a,b,decC)

File: codingChallenges/cpp/Codechef/longChallenges/Year2020/may/bitwise.cpp
#include<iostream>
#include<bits/stdc++.h>
#include<vector>
#define loop(n) for(int i=0;i<n;i++)
#define loop1(n) for(int i = n-1; i >= 0; i-- )
#define EXECUTE_FASTER ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
using namespace std;
#define ll long long int

void getBinary(ll n,vector<ll> & vect ){
    ll c =0,ans = 0;
    for(ll a=0,b=0,i=0;n>0;n=n/2,i++){
      a = n/2 ;
      b = n%2 ;
      vect.insert(vect.begin(),b);
    }
} 

ll getDecimal(vector<ll> & binC){
    ll n=0,c =0;
   for(ll i = binC.size()-1,b = 0; i >=0; i--,b++){
        n += binC[i]*pow(10,b);
    }
    for(ll a,i = 0;n > 0; i++,n = n/10){
        a =n%10 ;
        c =(pow(2,i))*a+c ; 
    }
    
    return c;
}

void cal(ll a,ll b,ll c){
    if ((a & c) == 0 or (b & c) == 0){
        cout << 0 << endl;
    }
    else{
        cout << c << endl;
    }  

}
void checkMaxCommon(vector<ll> binC,vector<ll> binR,int *k){
    int i;
    for(i = 0; binC[i] == binR[i]; i++){
    }
    k = &i;
}
int main(){
    EXECUTE_FASTER
    int t;
    cin >> t;
    ll a,b,l,r,c,d;
    while(t--){
        cin >> a >> b >> l >> r ;
        c = a | b;
        if( c <= r){
            cal(a,b,c);
            continue;
        }

        vector<ll> left,binR,binC,answer;
        getBinary(c,binC);
        getBinary(r,binR);
        int k;
        checkMaxCommon(binC,binR,&k);

        for(int i = 0; i < k; i++){
            answer.insert(answer.begin(),binR[i]);
        }
        answer.insert(answer.begin(),0);
        for(ll i = answer.size(); i < binC.size();i++){
            answer.push_back(binC[i]);
        }

        c = getDecimal(answer);
        if( c <= r){
            cal(a,b,c);
            continue;
        }




    }
    return 0;
}